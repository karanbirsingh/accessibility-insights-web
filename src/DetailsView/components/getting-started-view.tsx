// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { Assessment } from 'assessments/types/iassessment';
import { NamedFC } from 'common/react/named-fc';
import * as styles from 'DetailsView/components/getting-started-view.scss';
import {
    NextRequirementButton,
    NextRequirementButtonDeps,
} from 'DetailsView/components/next-requirement-button';
import * as React from 'react';
import { ContentLink, ContentLinkDeps } from 'views/content/content-link';

export type GettingStartedViewDeps = ContentLinkDeps & NextRequirementButtonDeps;
export interface GettingStartedViewProps {
    deps: GettingStartedViewDeps;
    assessment: Assessment;
}

export const GettingStartedView = NamedFC<GettingStartedViewProps>(
    'GettingStartedView',
    ({ deps, assessment }) => {
        const { gettingStarted, title, guidance, visualizationType, requirements } = assessment;

        const firstRequirement = requirements[0] ?? null;

        return (
            <div className={styles.gettingStartedView}>
                <div>
                    <h1 className={styles.gettingStartedHeader}>
                        <span className={styles.gettingStartedHeaderTitle}>{title}</span>
                        <ContentLink deps={deps} reference={guidance} iconName="info" />
                    </h1>
                    <h2 className={styles.gettingStartedTitle}>Getting started</h2>
                    {gettingStarted}
                </div>
                <NextRequirementButton
                    nextRequirement={firstRequirement}
                    currentTest={visualizationType}
                    className={styles.nextRequirementButton}
                    deps={deps}
                />
            </div>
        );
    },
);
