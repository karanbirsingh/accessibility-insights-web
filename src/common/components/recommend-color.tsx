// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

const MAXCOLORCOMPONENT: number = 255;
const MINCOLORCOMPONENT: number = 0;

export class RecommendColor {
    private _sentence: string;

    public getRecommendColor(fore: string, back: string, contrast: number): string {
        const foreRgb = this.hexToRGB(fore);
        const backRgb = this.hexToRGB(back);
        return (this._sentence = this.recommend(foreRgb, backRgb, contrast));
    }

    public get sentence(): string {
        return this._sentence;
    }

    private rgbToHex(r: number, g: number, b: number): string {
        return '#' + ((1 << 24) + (r << 16) + (g << 8) + b).toString(16).slice(1);
    }

    private hexToRGB(hex: string): number[] {
        const result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);
        if (result != null) {
            return [parseInt(result[1], 16), parseInt(result[2], 16), parseInt(result[3], 16)];
        }
        return [0, 0, 0];
    }

    private getLuminance(color: number[]): number {
        const a = [color[0], color[1], color[2]].map(function (v) {
            v /= 255;
            return v <= 0.03928 ? v / 12.92 : Math.pow((v + 0.055) / 1.055, 2.4);
        });
        return a[0] * 0.2126 + a[1] * 0.7152 + a[2] * 0.0722;
    }

    private darkenLighten(color: number[], shade: boolean): number[] {
        let num: number = 0.95;
        if (shade) {
            num = 1.05;
            if (color[0] === 0 && color[1] === 0 && color[2] === 0) {
                color = [10, 10, 10];
            }
        }

        for (let i = 0; i < 3; i++) {
            const component: number = color[i];
            if (component * num <= MAXCOLORCOMPONENT && component * num >= MINCOLORCOMPONENT) {
                color[i] = Math.floor(component * num);
            } else if (shade) {
                color[i] = MAXCOLORCOMPONENT;
            }
        }

        return color;
    }

    private recommendAColor(
        contrast: number,
        toBeChanged: number[],
        originalPair: number,
        toLightenDarken: boolean,
        colorContrast: number,
    ): object {
        let preContrast: number = -1;

        while (contrast < colorContrast && contrast !== preContrast) {
            preContrast = contrast;
            toBeChanged = this.darkenLighten(toBeChanged, toLightenDarken);
            if (toLightenDarken) {
                contrast = (this.getLuminance(toBeChanged) + 0.05) / (originalPair + 0.05);
            } else {
                contrast = (originalPair + 0.05) / (this.getLuminance(toBeChanged) + 0.05);
            }
        }

        return [toBeChanged, contrast];
    }

    private recommend(fore: number[], back: number[], colorContrast: number) {
        const foreOriginal: string = this.rgbToHex(fore[0], fore[1], fore[2]);
        const backOriginal: string = this.rgbToHex(back[0], back[1], back[2]);
        const foreLum: number = this.getLuminance(fore);
        const backLum: number = this.getLuminance(back);
        const lighter: number = Math.max(foreLum, backLum);
        const darker: number = lighter === foreLum ? backLum : foreLum;
        const contrast: number = (lighter + 0.05) / (darker + 0.05);

        let lightened: number[], darkened: number[];

        if (lighter === foreLum) {
            lightened = fore;
            darkened = back;
        } else {
            lightened = back;
            darkened = fore;
        }

        const recommendLightened: object = this.recommendAColor(
            contrast,
            lightened,
            darker,
            true,
            colorContrast,
        );
        const recommendDarkened: object = this.recommendAColor(
            contrast,
            darkened,
            lighter,
            false,
            colorContrast,
        );

        let sentence: string = ' ',
            foreRec: string,
            backRec: string;

        if (lighter === foreLum) {
            if (recommendLightened[1] > colorContrast) {
                foreRec = this.rgbToHex(
                    recommendLightened[0][0],
                    recommendLightened[0][1],
                    recommendLightened[0][2],
                );
                sentence = this.createSentence(
                    foreRec,
                    recommendLightened[1].toFixed(2),
                    sentence,
                    true,
                    backOriginal,
                );
            }
            if (recommendDarkened[1] > colorContrast) {
                backRec = this.rgbToHex(
                    recommendDarkened[0][0],
                    recommendDarkened[0][1],
                    recommendDarkened[0][2],
                );
                sentence += this.createSentence(
                    backRec,
                    recommendDarkened[1].toFixed(2),
                    sentence,
                    false,
                    foreOriginal,
                );
            }
        } else {
            if (recommendDarkened[1] > colorContrast) {
                foreRec = this.rgbToHex(
                    recommendDarkened[0][0],
                    recommendDarkened[0][1],
                    recommendDarkened[0][2],
                );
                sentence = this.createSentence(
                    foreRec,
                    recommendDarkened[1].toFixed(2),
                    sentence,
                    true,
                    backOriginal,
                );
            }
            if (recommendLightened[1] > colorContrast) {
                backRec = this.rgbToHex(
                    recommendLightened[0][0],
                    recommendLightened[0][1],
                    recommendLightened[0][2],
                );
                sentence += this.createSentence(
                    backRec,
                    recommendLightened[1].toFixed(2),
                    sentence,
                    false,
                    foreOriginal,
                );
            }
        }
        return sentence;
    }

    private createSentence(
        rec: string,
        contrast: number,
        sentence: string,
        foreBack: boolean,
        original: string,
    ) {
        if (foreBack) {
            sentence = ' Use foreground color: ' + rec + ' and the original background color: ';
        } else {
            if (sentence === ' ') {
                sentence += ' Use';
            } else {
                sentence = ' Or use';
            }
            sentence += ' background color: ' + rec + ' and the original foreground color: ';
        }
        sentence += original + ' to meet a contrast ratio of ' + contrast + ':1.';
        return sentence;
    }
}
