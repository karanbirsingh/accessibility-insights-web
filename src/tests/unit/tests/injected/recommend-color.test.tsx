// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { RecommendColor } from 'common/components/recommend-color';

describe('Recommend Color', () => {
    let testSubject: RecommendColor;

    beforeEach(() => {
        testSubject = new RecommendColor();
    });

    test('White on White recommendation', () => {
        testSubject.getRecommendColor('#ffffff', '#ffffff', 4.5);

        const result = testSubject.sentence;

        expect(result).toMatchSnapshot();
    });

    test('Black on black recommendation', () => {
        testSubject.getRecommendColor('#000000', '#000000', 4.5);

        const result = testSubject.sentence;

        expect(result).toMatchSnapshot();
    });

    test('Same Color recommendation', () => {
        testSubject.getRecommendColor('#8a94a8', '#8a94a8', 4.5);

        const result = testSubject.sentence;

        expect(result).toMatchSnapshot();
    });

    test('Normal Font Recommendation', () => {
        testSubject.getRecommendColor('#fefefe', '#21809d', 4.5);

        const result = testSubject.sentence;

        expect(result).toMatchSnapshot();
    });

    test('Large Font Recommendation', () => {
        testSubject.getRecommendColor('#0000D6', '#9A3743', 3.0);

        const result = testSubject.sentence;

        expect(result).toMatchSnapshot();
    });

    test('Incorrect Input', () => {
        testSubject.getRecommendColor('color', 'color', 4.5);

        const result = testSubject.sentence;

        expect(result).toMatchSnapshot();
    });
});
