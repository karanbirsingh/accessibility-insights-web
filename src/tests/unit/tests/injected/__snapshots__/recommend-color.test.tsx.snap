// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Recommend Color Black on black recommendation 1`] = `" "`;

exports[`Recommend Color Incorrect Input 1`] = `" "`;

exports[`Recommend Color Large Font Recommendation 1`] = `" Use foreground color: #000001 and the original background color: #9a3743 to meet a contrast ratio of 3.00:1. Or use background color: #f55266 and the original foreground color: #0000d6 to meet a contrast ratio of 3.19:1."`;

exports[`Recommend Color Normal Font Recommendation 1`] = `" Use foreground color: #ffffff and the original background color: #21809d to meet a contrast ratio of 4.53:1. Or use background color: #1f7995 and the original foreground color: #fefefe to meet a contrast ratio of 4.93:1."`;

exports[`Recommend Color Same Color recommendation 1`] = `"   Use background color: #292d33 and the original foreground color: #8a94a8 to meet a contrast ratio of 4.54:1."`;

exports[`Recommend Color White on White recommendation 1`] = `"   Use background color: #717171 and the original foreground color: #ffffff to meet a contrast ratio of 4.88:1."`;
